com.unity.shadergraph/Editor/Drawing/PreviewManager.cs

```diff
@@ -243,276 +282,443 @@ public bool HandleGraphChanges()
-                    renderData.shaderData.isCompiling = false;
+                    renderData.shaderData.passesCompiling = 0;
+                    renderData.shaderData.isOutOfDate = false;
                     CheckForErrors(renderData.shaderData);
-                    m_NodesToDraw.Add(renderData.shaderData.node);
+
+                    nodesCompiled.Add(renderData.shaderData.node);
 
                     var masterNode = renderData.shaderData.node as IMasterNode;
-                    masterNode?.ProcessPreviewMaterial(renderData.shaderData.mat);
+                    if (masterNode != null)
+                    {
+                        masterNode.ProcessPreviewMaterial(renderData.shaderData.mat);
+                    }
                 }
+
+                // removed compiled nodes from compiling list
+                m_NodesCompiling.ExceptWith(nodesCompiled);
+
+                // and add them to the draw list
+                m_NodesToDraw.UnionWith(nodesCompiled);
             }
+        }
 
-            if (m_NodesToUpdate.Count == 0)
-                return;
+        private static readonly ProfilerMarker KickOffShaderCompilationsMarker = new ProfilerMarker("KickOffShaderCompilations");
+        void KickOffShaderCompilations()
+        {
+            // Start compilation for nodes that need to recompile
+            using (KickOffShaderCompilationsMarker.Auto())
+            using (var nodesToCompile = PooledHashSet<AbstractMaterialNode>.Get())
+            {
+                // master node compile is first in the priority list, as it takes longer than the other previews
+                if ((m_NodesCompiling.Count + nodesToCompile.Count < m_MaxNodesCompiling) &&
+                    m_NodesNeedsRecompile.Contains(m_MasterRenderData.shaderData.node) &&
+                    !m_NodesCompiling.Contains(m_MasterRenderData.shaderData.node) &&
+                    ((Shader.globalRenderPipeline != null) && (Shader.globalRenderPipeline.Length > 0)))    // master node requires an SRP
+                {
+                    var renderData = GetPreviewRenderData(m_MasterRenderData.shaderData.node);
+                    Assert.IsTrue(renderData != null);
+                    nodesToCompile.Add(m_MasterRenderData.shaderData.node);
+                }
 
-            PropagateNodeList(m_NodesToUpdate, PropagationDirection.Downstream);
-            // Reset error states for the UI, the shader, and all render data for nodes we're updating
-            m_Messenger.ClearNodesFromProvider(this, m_NodesToUpdate);
-            var wasAsyncAllowed = ShaderUtil.allowAsyncCompilation;
-            ShaderUtil.allowAsyncCompilation = true;
+                // add each node to compile list if it needs a preview, is not already compiling, and we have room
+                // (we don't want to double kick compiles, so wait for the first one to get back before kicking another)
+                if (m_NodesCompiling.Count + nodesToCompile.Count < m_MaxNodesCompiling)
```

Comment:
If you move the `if(m_NodesCompiling.Count + ...) break;` (line 534) to the top of the for loop you can remove this if and simplify the logic.

Reply:
Yeah, I'll fix this.   Initializes the extra iterator, but that's not too bad, and the code reads better.